Could not initialize class java.awt.Toolkit

When running the following command `.\Jumpai.exe --cli --console --verbose` I get:

```
Changing working directory to C:\Users\Winter\Desktop\Jumpai-0.0.0-Windows ...
Loading JVM runtime library ...
Added DLL search directory C:\Users\Winter\Desktop\Jumpai-0.0.0-Windows\jre\bin
Added DLL search directory C:\Users\Winter\Desktop\Jumpai-0.0.0-Windows\jre\bin\server
Passing VM options ...
versionInformation.dwMajorVersion=10, versionInformation.dwMinorVersion=0, versionInformation.dwBuildNumber=19042
isZgcSupported()=1, hasJsonValue(jsonRoot, "useZgcIfSupportedOs", sajson::TYPE_TRUE)=0
versionInformation.dwMajorVersion=10, versionInformation.dwMinorVersion=0, versionInformation.dwBuildNumber=19042
Passing VM options:
Creating Java VM ...
Passing command line arguments ...
Loading JAR file ...
Adding 1 classpaths ...
  # Jumpai.data
Invoking static net.jumpai.client.desktop.DesktopLauncher.main() function ...
Checked for an exception from the main method, exceptionOccurred=1
Calling java.lang.Thread#dispatchUncaughtException(Throwable) on main thread
Exception in thread "main" java.lang.NoClassDefFoundError: Could not initialize class java.awt.Toolkit
        at java.desktop/java.awt.Component.<clinit>(Component.java:621)
        at net.jumpai.client.desktop.DesktopLauncher.main(DesktopLauncher.java:34)
Destroyed Java VM ...
Press ENTER key to exit.
```

I'm using AdoptOpenJDK 11 LTS with hotspot

My packr config is the following

```
{
    "platform": "windows64",
    "jdk": "out/jdks/win64-jdk",
    "executable": "Jumpai",
    "classpath": [
        "desktop/build/libs/Jumpai.data"
    ],
    "removelibs": [
        "desktop/build/libs/Jumpai.data"
    ],
    "mainclass": "net.jumpai.client.desktop.DesktopLauncher",
    "minimizejre": "none",
    "output": "out/packr/win64"
}
```

(I'm not minimizing the JRE).

This error is only happening on a fresh Windows 10 installation. On my old Windows 7, everything works fine. If I instead launch the game using `.\jre\bin\java.exe -jar Jumpai.data`, everything works fine. (This is using the JRE that was output by packr).

Any idea? Does packr gets rid of `java.awt` despite minimizejre being unspecified? If so, how come it works when using the same JRE and not the .exe?

I added the lines you mentionned and created my JFrame test, but gradle is unable to apply the plugin `org.gradle.cpp-application`

```
6:21:34 p.m.: Executing task 'createTestDirectory'...

> Task :buildSrc:compileKotlin UP-TO-DATE
> Task :buildSrc:compileJava NO-SOURCE
> Task :buildSrc:compileGroovy NO-SOURCE
> Task :buildSrc:pluginDescriptors UP-TO-DATE
> Task :buildSrc:processResources NO-SOURCE
> Task :buildSrc:classes UP-TO-DATE
> Task :buildSrc:inspectClassesForKotlinIC UP-TO-DATE
> Task :buildSrc:jar UP-TO-DATE
> Task :buildSrc:assemble UP-TO-DATE
> Task :buildSrc:compileTestKotlin NO-SOURCE
> Task :buildSrc:pluginUnderTestMetadata UP-TO-DATE
> Task :buildSrc:compileTestJava NO-SOURCE
> Task :buildSrc:compileTestGroovy NO-SOURCE
> Task :buildSrc:processTestResources NO-SOURCE
> Task :buildSrc:testClasses UP-TO-DATE
> Task :buildSrc:test NO-SOURCE
> Task :buildSrc:validatePlugins UP-TO-DATE
> Task :buildSrc:check UP-TO-DATE
> Task :buildSrc:build UP-TO-DATE
Configuration on demand is an incubating feature.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.6.1/userguide/command_line_interface.html#sec:command_line_warnings

FAILURE: Build failed with an exception.

* Where:
Build file '/home/winter/Packr/PackrLauncher/packrLauncher.gradle.kts' line: 27

* What went wrong:
An exception occurred applying plugin request [id: 'org.gradle.cpp-application']
> Failed to apply plugin 'org.gradle.cpp-application'.
   > Project#afterEvaluate(Action) on project ':PackrLauncher' cannot be executed in the current context.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 182ms
6:21:34 p.m.: Task execution finished 'createTestDirectory'.
```

Any clue on this one?

@karlsabo Same story on windows 7:

```
Alexander Winter@PC-Alex MINGW64 ~/Desktop/packr (master)
$ ./gradlew PackrAllTestApp:createTestDirectory
Starting a Gradle Daemon, 3 incompatible and 3 stopped Daemons could not be reused, use --status for details

> Task :buildSrc:jar
:jar: No valid plugin descriptors were found in META-INF/gradle-plugins
Configuration on demand is an incubating feature.

FAILURE: Build failed with an exception.

* Where:
Build file 'C:\Users\Alexander Winter\Desktop\packr\PackrLauncher\packrLauncher.gradle.kts' line: 27

* What went wrong:
An exception occurred applying plugin request [id: 'org.gradle.cpp-application']
> Failed to apply plugin 'org.gradle.cpp-application'.
   > Project#afterEvaluate(Action) on project ':PackrLauncher' cannot be executed in the current context.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.6.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 42s
```

This is from a fresh clone, I basically ran

```
git clone http://github.com/libgdx/packr
cd packr
./gradlew PackrAllTestApp:createTestDirectory
```

This issue might be bothersome but that's really not the actual issue. @karlsabo Can you just send me a zip of your test app so I can test it on my fresh windows 10?

--- a/Packr/packr.gradle.kts
+++ b/Packr/packr.gradle.kts
@@ -55,8 +55,8 @@ repositories {
    gitHubRepositoryForPackr(project)
 
    // temporary for CI publishing until oss.sonatype.org is available for com.libgdx.packr or com.badlogicgames.packr
-   maven("http://artifactory.nimblygames.com/artifactory/ng-public-snapshot/")
-   maven("http://artifactory.nimblygames.com/artifactory/ng-public-release/")
+   maven("https://artifactory.nimblygames.com/artifactory/ng-public-snapshot/")
+   maven("https://artifactory.nimblygames.com/artifactory/ng-public-release/")
 }
 
 java {
@@ -264,14 +264,14 @@ publishing {
       if (ngToken != null) {
          val ngUsername = findProperty("NG_ARTIFACT_REPOSITORY_USER") as String? ?: System.getenv("NG_ARTIFACT_REPOSITORY_USER")
          if (isSnapshot) {
-            maven("http://artifactory.nimblygames.com/artifactory/ng-public-snapshot/") {
+            maven("https://artifactory.nimblygames.com/artifactory/ng-public-snapshot/") {
                credentials {
                   username = ngUsername
                   password = ngToken
                }
             }
          } else {
-            maven("http://artifactory.nimblygames.com/artifactory/ng-public-release/") {
+            maven("https://artifactory.nimblygames.com/artifactory/ng-public-release/") {
                credentials {
                   username = ngUsername
                   password = ngToken

--- a/PackrAllTestApp/packrAllTestApp.gradle.kts
+++ b/PackrAllTestApp/packrAllTestApp.gradle.kts
@@ -216,6 +216,9 @@ val createTestDirectory: TaskProvider<Task> = tasks.register("createTestDirector
             if (!outputAsString.contains("Received uncaught exception")) {
                throw GradleException("Packr bundle in $packrOutputDirectory didn't catch an unchecked exception with setDefaultUncaughtExceptionHandler")
             }
+            if (!outputAsString.contains("Testing uncaught exception handler.")) {
+               throw GradleException("Packr bundle in $packrOutputDirectory didn't throw \"Testing uncaught exception handler.\"")
+            }
             if (fileNameNoExtension.toLowerCase()
                    .contains("jdk14") && !outputAsString.contains("Using The Z Garbage Collector")) {
                throw GradleException("Packr bundle in $packrOutputDirectory didn't execute using the Z garbage collector")

--- a/PackrAllTestApp/packrAllTestApp.gradle.kts
+++ b/PackrAllTestApp/packrAllTestApp.gradle.kts
@@ -309,6 +309,8 @@ fun createPackrContent(jdkPath: Path, osFamily: String, destination: Path) {
       args("--vmargs")
       args("Dsun.java2d.noddraw=true")
       args("--vmargs")
+      args("Djava.awt.headless=true")
+      args("--vmargs")
       args("XstartOnFirstThread")
       if (jdkPath.fileName.toString().toLowerCase().contains("jdk14")) {
          args("--useZgcIfSupportedOs")

--- a/PackrAllTestApp/src/main/java/com/badlogicgames/packrtestapp/PackrAllTestApplication.java
+++ b/PackrAllTestApp/src/main/java/com/badlogicgames/packrtestapp/PackrAllTestApplication.java
@@ -16,10 +16,13 @@
 
 package com.badlogicgames.packrtestapp;
 
+import javax.swing.JFrame;
+import javax.swing.JLabel;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.stream.Stream;
 
 /**
  * Simple hello world application for testing the packr launcher.
@@ -33,6 +36,13 @@ public class PackrAllTestApplication {
 	  * @throws IOException if an IO error occurs
 	  */
 	 public static void main (String[] args) throws IOException, InterruptedException {
+		  JFrame myFrame = new JFrame("Hello");
+		  myFrame.getContentPane().add(new JLabel("Hello world!"));
+		  myFrame.pack();
+		  myFrame.setLocationRelativeTo(null);
+		  myFrame.setVisible(false);
+		  myFrame.dispose();
+
 		  System.out.println("Hello world!");
 		  System.out.println("Running from java.version=" + System.getProperty("java.version"));
 
@@ -46,8 +56,9 @@ public class PackrAllTestApplication {
 				throwable.printStackTrace(System.out);
 		  });
 
-		  Files.lines(Paths.get("application-resources").resolve("fake-resource.txt"))
-			  .forEach(resourceLine -> System.out.println("Loaded resource line: " + resourceLine));
+		  try (Stream<String> lineStream = Files.lines(Paths.get("application-resources").resolve("fake-resource.txt"))) {
+				lineStream.forEach(resourceLine -> System.out.println("Loaded resource line: " + resourceLine));
+		  }
 
 		  throw new RuntimeException(
 			  "Testing uncaught exception handler. Thrown from the main thread, Thread.currentThread().getName()=" + Thread.currentThread().getName());

--- a/PackrAllTestApp/src/main/java/com/badlogicgames/packrtestapp/PackrAllTestApplication.java
+++ b/PackrAllTestApp/src/main/java/com/badlogicgames/packrtestapp/PackrAllTestApplication.java
@@ -35,7 +35,7 @@ public class PackrAllTestApplication {
 	  * @throws IOException if an IO error occurs
 	  */
 	 public static void main (String[] args) throws IOException, InterruptedException {
-		  System.out.println("ScreenSize=" + Toolkit.getDefaultToolkit().getScreenSize());
+		  System.out.println("EventQueue=" + Toolkit.getDefaultToolkit().getSystemEventQueue());
 
 		  System.out.println("Hello world!");
 		  System.out.println("Running from java.version=" + System.getProperty("java.version"));

--- a/PackrAllTestApp/src/main/java/com/badlogicgames/packrtestapp/PackrAllTestApplication.java
+++ b/PackrAllTestApp/src/main/java/com/badlogicgames/packrtestapp/PackrAllTestApplication.java
@@ -16,8 +16,7 @@
 
 package com.badlogicgames.packrtestapp;
 
-import javax.swing.JFrame;
-import javax.swing.JLabel;
+import java.awt.Toolkit;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -36,12 +35,7 @@ public class PackrAllTestApplication {
 	  * @throws IOException if an IO error occurs
 	  */
 	 public static void main (String[] args) throws IOException, InterruptedException {
-		  JFrame myFrame = new JFrame("Hello");
-		  myFrame.getContentPane().add(new JLabel("Hello world!"));
-		  myFrame.pack();
-		  myFrame.setLocationRelativeTo(null);
-		  myFrame.setVisible(false);
-		  myFrame.dispose();
+		  System.out.println("ScreenSize=" + Toolkit.getDefaultToolkit().getScreenSize());
 
 		  System.out.println("Hello world!");
 		  System.out.println("Running from java.version=" + System.getProperty("java.version"));
@@ -50,8 +44,8 @@ public class PackrAllTestApplication {
 		  new ProcessBuilder(javaExecutablePath.toString(), "-version").inheritIO().start().waitFor();
 
 		  Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
-				System.out.println(
-					"Received uncaught exception in thread.getName()=" + thread.getName() + ", Thread.currentThread().getName()=" + Thread.currentThread()
+				System.out
+					.println("Received uncaught exception in thread.getName()=" + thread.getName() + ", Thread.currentThread().getName()=" + Thread.currentThread()
 						.getName());
 				throwable.printStackTrace(System.out);
 		  });

--- a/PackrLauncher/packrLauncher.gradle.kts
+++ b/PackrLauncher/packrLauncher.gradle.kts
@@ -299,14 +299,14 @@ publishing {
       if (ngToken != null) {
          val ngUsername = findProperty("NG_ARTIFACT_REPOSITORY_USER") as String? ?: System.getenv("NG_ARTIFACT_REPOSITORY_USER")
          if (isSnapshot) {
-            maven("http://artifactory.nimblygames.com/artifactory/ng-public-snapshot/") {
+            maven("https://artifactory.nimblygames.com/artifactory/ng-public-snapshot/") {
                credentials {
                   username = ngUsername
                   password = ngToken
                }
             }
          } else {
-            maven("http://artifactory.nimblygames.com/artifactory/ng-public-release/") {
+            maven("https://artifactory.nimblygames.com/artifactory/ng-public-release/") {
                credentials {
                   username = ngUsername
                   password = ngToken

--- a/PackrLauncher/src/main/cpp/win32/packr_win32.cpp
+++ b/PackrLauncher/src/main/cpp/win32/packr_win32.cpp
@@ -251,9 +251,43 @@ void addDllDirectory(LPCWSTR directory) {
    }
 }
 
+/**
+ * Loads every library found using the pattern {@code libraryPattern}.
+ * @param libraryPattern the search pattern to use to locate shared libaries to load
+ */
+void loadLibraries(const PTCHAR libraryPattern) {
+   WIN32_FIND_DATA FindFileData;
+   HANDLE hFind = nullptr;
+   TCHAR libraryPath[MAX_PATH];
+
+   wstring_convert<codecvt_utf8_utf16<wchar_t>> converter;
+
+   hFind = FindFirstFile(libraryPattern, &FindFileData);
+   if (hFind == INVALID_HANDLE_VALUE) {
+      if (verbose) {
+         cout << "Couldn't find " << converter.to_bytes(libraryPattern) << " file." << "FindFirstFile failed " << GetLastError() << "." << endl;
+      }
+      return;
+   }
+   do {
+      if (LoadLibraryEx(FindFileData.cFileName, nullptr, LOAD_LIBRARY_SEARCH_DEFAULT_DIRS) == nullptr) {
+         if (verbose) {
+            cout << "Failed to load DLL " << converter.to_bytes(FindFileData.cFileName) << "." << endl;
+         }
+      } else if (verbose) {
+         cout << "Loaded DLL " << converter.to_bytes(FindFileData.cFileName) << "." << endl;
+      }
+   } while (FindNextFile(hFind, &FindFileData));
+
+   FindClose(hFind);
+}
+
 bool loadJNIFunctions(GetDefaultJavaVMInitArgs *getDefaultJavaVMInitArgs, CreateJavaVM *createJavaVM) {
    addDllDirectory(TEXT("jre\\bin"));
    addDllDirectory(TEXT("jre\\bin\\server"));
+
+   // Load every shared library in jre/bin because awt.dll doesn't load its dependent libraries using the correct search paths
+   loadLibraries(TEXT("jre\\bin\\*.dll"));
 
    TCHAR jvmDllFullPath[FULL_PATH_SIZE] = TEXT("");
    if (GetFullPathName(TEXT("jre\\bin\\server\\jvm.dll"), FULL_PATH_SIZE, jvmDllFullPath, nullptr) == 0) {

--- a/TestAppJreDist/testAppJreDist.gradle.kts
+++ b/TestAppJreDist/testAppJreDist.gradle.kts
@@ -332,6 +332,8 @@ extractJdkTasks.forEach { extractJdkTask ->
             args(jlinkPath.toAbsolutePath().toString())
             args("--add-modules")
             args("java.base")
+            args("--add-modules")
+            args("java.desktop")
          }
       }
    }

--- a/changelog.md
+++ b/changelog.md
@@ -1,3 +1,6 @@
+# Unreleased 3.0.2
+1. Preemptively load all DLLs in the jre/bin directory on Windows.
+   * This resolves an issue on Windows where the awt.dll cannot find the Microsoft runtime libraries (MSVCP140.dll).
 # Release 3.0.1
 1. Updated tests to use the latest versions of AdoptOpenJDK 8, 11, and 15.
 1. Changed DLL loading on Windows to search in the JRE bin and server directories.

